src/Control/Monad/Log.hs

```diff
@@ -491,8 +491,11 @@ withBatchedHandler BatchingOptions{..} flush k =
                        flushAfter flushAlarm <|> flushFull <|> flushOnClose
                      stillOpen <- fmap not (readTVar closed)
                      return (messages,stillOpen))
-             flush messages
-             pure stillOpen
+             case messages of
+                [] -> pure stillOpen
+                msgs ->
+                    do flush msgs
```

Comment:
This looks good, but I think the type of `flush` should change. We only ever call ` flush` with a non-empty list, so we should use `Data.List.NonEmpty`.